lib/tty/pager/basic.rb

```diff
@@ -43,48 +45,112 @@ def default_prompt
       #
       # @api public
       def page(text, &callback)
-        page_num = 1
-        leftover = []
-        lines_left = @height
+        write(text, &callback)
+      rescue PagerClosed
+        # do nothing
+      ensure
+        reset
+      end
+
+      # Write text to the pager, prompting on page end.
+      #
+      # @raise [PagerClosed]
+      #   if the pager was closed
+      #
+      # @return [TTY::Pager::BasicPager]
+      #
+      # @api public
+      def write(*args, &callback)
+        args.each do |text|
+          send_text(:write, text, &callback)
+        end
+        self
+      end
+      alias_method :<<, :write
+
+      # Write text to the pager, prompting on page end. Returns false if the
+      # pager was closed.
+      #
+      # @return [Boolean]
+      #   the success status of writing to the screen
+      #
+      # @api public
+      def try_write(*args, &callback)
+        write(text, &callback)
+        true
+      rescue PagerClosed
+        false
+      end
 
+      # Print a line of text to the pager, prompting on page end.
+      #
+      # @raise [PagerClosed]
+      #   if the pager was closed
+      #
+      # @api public
+      def puts(text, &callback)
+        send_text(:puts, text, &callback)
+      end
+
+      # The lower-level common implementation of printing methods
+      #
+      # @return [Boolean]
+      #   the success status of writing to the screen
+      #
+      # @api private
+      def send_text(write_method, text, &callback)
         text.lines.each do |line|
           chunk = []
-          if !leftover.empty?
-            chunk = leftover
-            leftover = []
+          if !@leftover.empty?
+            chunk = @leftover
+            @leftover = []
           end
           wrapped_line = Strings.wrap(line, @width)
           wrapped_line.lines.each do |line_part|
-            if lines_left > 0
+            if @lines_left > 0
               chunk << line_part
-              lines_left -= 1
+              @lines_left -= 1
             else
-              leftover << line_part
+              @leftover << line_part
             end
           end
-          output.print(chunk.join)
+          output.public_send(write_method, chunk.join)
 
-          if lines_left == 0
-            break unless continue_paging?(page_num)
-            lines_left = @height
-            if leftover.size > 0
-              lines_left -= leftover.size
+          if @lines_left == 0
+            raise PagerClosed unless continue_paging?(@page_num)
```

Comment:
I've added a more readable error message, and I've made it specific to the pager, with more information for debugging purposes. Let me know what you think.

Regarding the "closed" variable, I'm not sure how it would work to avoid the `try_write`. I'm imagining something like

``` ruby
if !pager.closed?
  pager.write(...)
end
```

But we can't have an instance variable for that -- the act of writing is the thing that discovers that the stream is closed, and that happens after the check. We could check the `closed?` state of the pipe, but then there could be a timing issue with the check succeeding, but then in the middle of the `write` call the user closes the pager. It's probably unlikely, it just feels a bit flaky to me. Let me know if I'm misunderstanding the scenario, though.

Reply:
I agree that's probably not a substitute for the `try_wite` in which case it makes sense to keep it. The `closed?` relates more to the process/pipe dying. When the paging tool died then 100% you cannot make any writes. And this probably works better with `<<` to provide an extra layer of robustness. However, this won't mitigate the scenario when the failure happens during the write - then all bets are off. In which case raising an error is a reasonable approach or calling `try_write`.